File: c#/BigEgg.Algorithm/BigEgg.Algorithm.Test/UnionFinds/QuickUnionUFTest.cs
using BigEgg.Algorithm.UnionFinds;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;

namespace BigEgg.Algorithm.Test.UnionFinds
{
    [TestClass]
    public class QuickUnionUFTest
    {
        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testUnion_ParameterValidation_PLessThan0()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Union(-1, 4);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testUnion_ParameterValidation_PEqualN()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Union(10, 4);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testUnion_ParameterValidation_PLargerThanN()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Union(11, 4);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testUnion_ParameterValidation_QLessThan0()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Union(4, -1);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testUnion_ParameterValidation_QEqualN()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Union(4, 10);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testUnion_ParameterValidation_QLargerThanN()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Union(4, 11);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testConnected_ParameterValidation_PLessThan0()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Connected(-1, 4);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testConnected_ParameterValidation_PEqualN()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Connected(10, 4);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testConnected_ParameterValidation_PLargerThanN()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Connected(11, 4);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testConnected_ParameterValidation_QLessThan0()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Connected(4, -1);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testConnected_ParameterValidation_QEqualN()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Connected(4, 10);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testConnected_ParameterValidation_QLargerThanN()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Connected(4, 11);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testFind_ParameterValidation_PLessThan0()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Find(-1);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testFind_ParameterValidation_PEqualN()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Find(10);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void testFind_ParameterValidation_PLargerThanN()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Find(11);
        }


        [TestMethod]
        public void testUnionAndConnected_NewConnection()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            Assert.IsFalse(quickUnionUF.Connected(0, 1));
            Assert.AreEqual(10, quickUnionUF.Count());
            quickUnionUF.Union(0, 1);
            Assert.IsTrue(quickUnionUF.Connected(0, 1));
            Assert.AreEqual(9, quickUnionUF.Count());
            Assert.AreEqual(1, quickUnionUF.Find(1));
            Assert.AreEqual(1, quickUnionUF.Find(0));
        }

        [TestMethod]
        public void testUnionAndConnected_AlreadyConnected()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Union(0, 1);
            Assert.IsTrue(quickUnionUF.Connected(0, 1));
            Assert.AreEqual(9, quickUnionUF.Count());
            quickUnionUF.Union(0, 1);
            Assert.IsTrue(quickUnionUF.Connected(0, 1));
            Assert.AreEqual(9, quickUnionUF.Count());
            Assert.AreEqual(1, quickUnionUF.Find(1));
            Assert.AreEqual(1, quickUnionUF.Find(0));
        }

        [TestMethod]
        public void testUnionAndConnected_MultiConnection()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            quickUnionUF.Union(0, 1);
            quickUnionUF.Union(2, 1);
            Assert.IsTrue(quickUnionUF.Connected(0, 1));
            Assert.IsTrue(quickUnionUF.Connected(2, 1));
            Assert.IsTrue(quickUnionUF.Connected(2, 0));
            Assert.AreEqual(8, quickUnionUF.Count());
            Assert.AreEqual(1, quickUnionUF.Find(1));
            Assert.AreEqual(1, quickUnionUF.Find(0));
            Assert.AreEqual(1, quickUnionUF.Find(2));

            quickUnionUF.Union(2, 5);
            Assert.IsTrue(quickUnionUF.Connected(0, 1));
            Assert.IsTrue(quickUnionUF.Connected(2, 1));
            Assert.IsTrue(quickUnionUF.Connected(2, 0));
            Assert.IsTrue(quickUnionUF.Connected(5, 0));
            Assert.IsTrue(quickUnionUF.Connected(5, 1));
            Assert.IsTrue(quickUnionUF.Connected(5, 2));
            Assert.AreEqual(7, quickUnionUF.Count());
            Assert.AreEqual(5, quickUnionUF.Find(1));
            Assert.AreEqual(5, quickUnionUF.Find(0));
            Assert.AreEqual(5, quickUnionUF.Find(2));
            Assert.AreEqual(5, quickUnionUF.Find(5));
        }

        [TestMethod]
        public void testFind()
        {
            QuickUnionUF quickUnionUF = new QuickUnionUF(10);
            for (int i = 0; i < 10; i++)
            {
                Assert.AreEqual(i, quickUnionUF.Find(i));
            }
        }
    }
}

File: java/Algorithms/test/UnionFinds/QuickUnionUFTest.java
package UnionFinds;

import org.junit.Test;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

public class QuickUnionUFTest {
    @Test(expected = IndexOutOfBoundsException.class)
    public void testUnion_ParameterValidation_PLessThan0() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.union(-1, 4);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testUnion_ParameterValidation_PEqualN() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.union(10, 4);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testUnion_ParameterValidation_PLargerThanN() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.union(11, 4);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testUnion_ParameterValidation_QLessThan0() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.union(4, -1);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testUnion_ParameterValidation_QEqualN() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.union(4, 10);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testUnion_ParameterValidation_QLargerThanN() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.union(4, 11);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testConnected_ParameterValidation_PLessThan0() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.connected(-1, 4);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testConnected_ParameterValidation_PEqualN() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.connected(10, 4);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testConnected_ParameterValidation_PLargerThanN() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.connected(11, 4);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testConnected_ParameterValidation_QLessThan0() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.connected(4, -1);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testConnected_ParameterValidation_QEqualN() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.connected(4, 10);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testConnected_ParameterValidation_QLargerThanN() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.connected(4, 11);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testFind_ParameterValidation_PLessThan0() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.find(-1);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testFind_ParameterValidation_PEqualN() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.find(10);
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testFind_ParameterValidation_PLargerThanN() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.find(11);
    }


    @Test
    public void testUnionAndConnected_NewConnection() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        assertThat(quickUnionUF.connected(0, 1), is(false));
        assertThat(quickUnionUF.count(), is(10));
        quickUnionUF.union(0, 1);
        assertThat(quickUnionUF.connected(0, 1), is(true));
        assertThat(quickUnionUF.count(), is(9));
        assertThat(quickUnionUF.find(1), is(1));
        assertThat(quickUnionUF.find(0), is(1));
    }

    @Test
    public void testUnionAndConnected_AlreadyConnected() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.union(0, 1);
        assertThat(quickUnionUF.connected(0, 1), is(true));
        assertThat(quickUnionUF.count(), is(9));
        quickUnionUF.union(0, 1);
        assertThat(quickUnionUF.connected(0, 1), is(true));
        assertThat(quickUnionUF.count(), is(9));
        assertThat(quickUnionUF.find(1), is(1));
        assertThat(quickUnionUF.find(0), is(1));
    }

    @Test
    public void testUnionAndConnected_MultiConnection() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        quickUnionUF.union(0, 1);
        quickUnionUF.union(2, 1);
        assertThat(quickUnionUF.connected(0, 1), is(true));
        assertThat(quickUnionUF.connected(2, 1), is(true));
        assertThat(quickUnionUF.connected(2, 0), is(true));
        assertThat(quickUnionUF.count(), is(8));
        assertThat(quickUnionUF.find(1), is(1));
        assertThat(quickUnionUF.find(0), is(1));
        assertThat(quickUnionUF.find(2), is(1));

        quickUnionUF.union(2, 5);
        assertThat(quickUnionUF.connected(0, 1), is(true));
        assertThat(quickUnionUF.connected(2, 1), is(true));
        assertThat(quickUnionUF.connected(2, 0), is(true));
        assertThat(quickUnionUF.connected(5, 0), is(true));
        assertThat(quickUnionUF.connected(5, 1), is(true));
        assertThat(quickUnionUF.connected(5, 2), is(true));
        assertThat(quickUnionUF.count(), is(7));
        assertThat(quickUnionUF.find(1), is(5));
        assertThat(quickUnionUF.find(0), is(5));
        assertThat(quickUnionUF.find(2), is(5));
        assertThat(quickUnionUF.find(5), is(5));
    }

    @Test
    public void testFind() throws Exception {
        QuickUnionUF quickUnionUF = new QuickUnionUF(10);
        for (int i = 0; i < 10; i++) {
            assertThat(quickUnionUF.find(i), is(i));
        }
    }
}